docs/developer/designs/hierarchical-podgroup/README.md

```diff
@@ -0,0 +1,263 @@
+
+## Goals
+- Enable fine-grained scheduling specification: Allow PodGroups to define how smaller SubGroups of pods should be scheduled, specifying minMember and topology constraints per SubGroup.
+- Preserve atomicity while enabling flexibility: Maintain gang scheduling semantics at the workload level while enabling per-SubGroup constraints for granular control over subsets of pods.
+- Support advanced topology-aware placement: Allow SubGroups within a PodGroup to define specific topology constraints for precise placement within the cluster.
+- Enhance workload-aligned resource utilization: Improve cluster utilization by supporting differentiated placement strategies without imposing overly restrictive uniform constraints.
+- Maintain backward compatibility: Ensure existing workloads using PodGroups without SubGroups continue to function without modification.
+
+## Non-Goals
+- Partial Workload Execution: This design does not support executing subsets of pods independently before the PodGroup’s gang scheduling conditions are met.
+- Topology Discovery and Optimization: The design does not include enhancements for topology discovery or automated topology optimization.
+- Changes to Pod-Level Scheduling: This enhancement does not alter pod-level scheduling mechanisms beyond the SubGroup minMember and topology constraints applied within the PodGroup context.
+
+## User Stories
+
+### User Story 1: Fine-Grained Gang Scheduling
+I want to specify a PodGroup requiring 10 pods to be scheduled atomically, while defining that the decode-workers SubGroup requires at least 4 pods to be scheduled as a gang, enabling better resource alignment for my workload’s architecture.
+
+### User Story 2: SubGroup-Specific Topology Constraints
+I want one SubGroup to be placed within the same rack for low latency, while allowing another SubGroup pods to be placed flexibly across the cluster, aligning network-sensitive components with workload requirements.
+
+### User Story 3: Multi-SubGroup Workload with Co-Located Leaders and Workers
+I want to define decode-leaders, decode-workers, prefill-leaders, and prefill-workers SubGroups where leaders and workers are co-located within each group, but prefill and decode groups may be scheduled in different locations to optimize resource placement.
+
+### User Story 4: Backward Compatibility
+I want my existing PodGroups without SubGroups to continue working without modification, ensuring a seamless migration path for advanced scheduling capabilities.
+
```

Comment:
I would also add a user story for "Elastic subgroups" - To allocate the Podgroup, we need to be able to schdule some of the subgroups, but not all. For example, We would like to schedule a leaderWorkerSet object only if we can schedule at least 3 out of 5 of its replicas.